Input history

arl opened the pull request.

arl committed:
changed qlineedit to qcombobox

Conflicts:
	mainwindow.cpp

intermediary commit

add historymodel

for now historymodel has basic functionnalities only

delete historymodel

arl committed:
add HistoryComboBox and History classes

lmancini commented on history.cpp:
```diff
@@ -0,0 +1,68 @@
+    // remove eventual duplicates of new entry
+    int dup = history.indexOf(entry);
+    if (dup >= 0)
+        history.removeAt(dup);
```
Ti faccio solo notare che parli di `duplicateS` quando in realtà, per costruzione, ce ne sarà al più uno soltanto: l'invariante non scritta di questa classe infatti è che i duplicati non entrano perché sono filtrati proprio da questo codice, che è anche l'unico modo di aggiungere entry.

arl replied:
Infatti, non si puo avere piu di una copia di una string, e non si puo avere neanche piu di una string con altre copie, perché aggiungiamo string una alla volta.
Ho appena fatto un commit per questo

arl committed:
comment modifications

lmancini commented on historycombobox.cpp:
```diff
@@ -0,0 +1,67 @@
+void HistoryComboBox::fillList(QString current_text)
+{
+    QStringList hist_lines(history->getHistory());
+    std::reverse(hist_lines.begin(), hist_lines.end());
```
Apprezzo l'uso di `<algorithm>`, ma leggendo questo codice mi sembra che l'unica volta che accedi all'istanza di `History` come prima cosa ne inverti gli elementi: forse ti conviene memorizzarli direttamente nell'ordine inverso?

arl replied:
Si hai totalemente raggione, possiamo togliere il `std::reverse` call, fatto ad ogni modifica della lista, se gestiamo la lista in ordine reverse dall inizio.

arl committed:
reverse order of history list

as seen with Lorenzo, this saves us a full-rewrite of the list, at each
list modification

aurelien-rainone commented:
Ho appena visto qualcosa altro. 
Se apri la listbox dell history combo box, e selezionni (right-click mouse) un entry. Il cursor interno dell history non e aggiornato al nuovo elemento selezionato

arl committed:
manual update history pointer from combobox::activated

arl committed:
add code to detect if we can or not go up the history list

arl committed:
set max history entries to 10

lmancini commented on historycombobox.cpp:
```diff
@@ -0,0 +1,74 @@
+            setItemText(0, history->next());
+            fillList(this->lineEdit()->text());
+            break;
+
```
Credo che questi due casi qua sopra (`Qt::Key_Up`e `Qt::Key_Down`) debbano essere invertiti: nella shell, se non stai navigando la history e premi la freccia "up", ti appare il comando più recente, mentre col codice che hai scritto non succede niente (perché siamo in fondo alla coda della history).

Può essere che quando hai invertito l'ordinamento della lista della history tu dovessi passare anche di qui per completare la modifica?

lmancini commented:
A parte il mio ultimo commento la PR mi sembra ok.

arl committed:
inverse history behaviour (display and arrows)

Before, pressing the down arrow key meant displaying previous element in
history list. Now this behaviour is reversed. To do this new elements
are  appended at the front of the underlying history list

aurelien-rainone commented:
roles of up and down keys have been switched

arl committed:
useless parenthesis
